### ISSUE TITLE
Implement Smart Folders for Note Organization

### ISSUE THREAD

**Issue by SankethBK** [SankethBK/diaryvault#219]
**Description:**

Currently, we don't have a feature to organize notes in the form of folders. Since we have support for adding tags to notes, we can leverage this to build support for **Smart Folders**. The idea is that users can create a virtual folder and select a list of tags; all notes having these tags will automatically appear in these folders.

**Note:** Since this is a large task, you can contribute to a subset of the tasks outlined below.

---

### Data Source Changes

- [ ] **Create a New Table: `smart_folders`**
  - In `lib/core/databases/sqflite_setup.dart`, create a new table named `smart_folders`.
  - Increase the database version to **3**.
  - Add the create table query in the `onUpgrade` method.

- [ ] **Set Up Models and Entities**
  - Create a new folder `smart_folders` in `lib/features/`.
  - Define a model and entity to represent Smart Folders.

- [ ] **Define the Data Source Interface**
  - In `lib/features/smart_folders/data/datasources/local_data_sources/`, define an abstract class `ISmartFoldersLocalDataSource`.
  - Define contracts for CRUD operations related to smart folders.

- [ ] **Implement the Local Data Source**
  - Implement the methods defined in `ISmartFoldersLocalDataSource` in a class `SmartFoldersLocalDataSource`.
  - This involves writing SQL queries to interact with the smart folders table.

### Presentation Layer Changes

- [ ] **Add Smart Folders Section to Home Page**
  - Create a `PageView` in the home page and include a Smart Folders section.
  - **Mockup:**
     <img src="https://github.com/user-attachments/assets/e5b7715d-07c5-4010-b685-f95273380ed4" width="500" />

- [ ] **Implement Blocs and Cubits**
  - Create the necessary blocs and cubits to interact with the Smart Folders data source APIs.

- [ ] **Design the Smart Folders List Page**
  - Create the UI for the Smart Folders list page (similar to the notes list page).
  - **Mockup:** 
   <img src="https://github.com/user-attachments/assets/fa2db904-29db-4969-837f-e4bcdde5dfa7" width="500" />
- [ ] **Create Smart Folder Creation Popup**
  - Implement a UI for a Smart Folder creation popup that opens when clicking on the FAB (Floating Action Button).
  - The popup should allow the user to input:
    - **Smart Folder Title**
    - **List of Tags**
  - UI and APIs related to tags can be reused.
  - **Mockup:** *(To be created)*

---

**Additional Notes:**

- Ensure that the implementation aligns with the existing architecture and coding standards.
- Reuse existing components where possible to maintain consistency.
- Consider the impact on performance when fetching and displaying notes associated with Smart Folders.
- If you have any questions or need clarification, feel free to reach out to the maintainers.

---

**References:**

- **Existing Widgets and Files:**
  - `lib/core/databases/sqflite_setup.dart`
  - `lib/features/notes/...` (for reference on models and data sources)
  - `lib/features/tags/...` (for reusing tag-related UI and APIs)

---

Feel free to contribute to any of the tasks listed above. Your contributions are highly appreciated!

**Note:** Please comment on this issue to let others know which task you're working on to avoid duplication of effort.

**Comment by Lucie4914** [comment]
Hello, I am interested in contributing to this but it would my first time. What do you recommend I start with? I have some familiarity with Flutter from a past project though I am more at ease with SQL.

**Comment by Lucie4914** [comment]
I am not sure if you get notified without a mention, so I'll mention you just in case @SankethBK ^

**Comment by SankethBK** [comment]
Hi @Lucie4914 , sorry missed your previous comment. This is a bit larger feature, if you're interested you can start with creating database tables and entity models

**Comment by Lucie4914** [comment]
Alright, I'm currently taking a look at the code to create the table but before modifying anything I wanted to run the tests to verify my installation was correct and I found that some tests currently don't pass. Is it a problem on my side or not? Here are the test logs:

[test_results.log](https://github.com/user-attachments/files/23261046/test_results.log)

**Comment by SankethBK** [comment]
Test cases are quite old so I guess they'll be outdated. Don't worry about breaking, I'll review the changes

**Comment by Lucie4914** [comment]
Should the smart folder table contain the author ID ? Currently I've made it so it doesn't but I'm not sure if it's required for other features.
Currently I have the following columns: folder ID, folder name, tags, createdAt

**Comment by SankethBK** [comment]
Yeah I think better to add it , to keep it consistent with notes. It will help to support multiple users on same device

**Comment by Lucie4914** [comment]
I'm doing the SQL queries now and I realised I don't have a `deleted` attribute for the SmartFolders in case the hard delete is set to false. Is that a big deal? I feel like it's not needed here but I prefer asking to double check.

Also, I'm a bit confused on why `authorId` is a required argument for almost all the SQL queries when it's sometimes not used at all, such as for `getNote` and `deleteNote`. Should I filter smart folders by authorId even if it's a bit inconsistent with notes? (In my opinion yes, but I prefer to ask)

**Comment by SankethBK** [comment]
It's fine, I don't think it's needed.

**Comment by Lucie4914** [comment]
I've created a [pull request](https://github.com/SankethBK/diaryvault/pull/310) to merge all the data source changes
